applications/NeuralNetworkApplication/python_scripts/data_generator_process.py

```diff
@@ -0,0 +1,361 @@
+                self.hdf5_input_parameters["list_of_variables"].SetStringArray(input_var_names.GetStringArray())
+                self.hdf5_input_parameters.AddEmptyValue("prefix")
+                #self.hdf5_input_parameters["prefix"].SetString("/<time>/<model_part_name>/InputData")               
+
+                # Training output
+                training_output_file_name = settings["training_file_settings"]["output_path"].GetString() + training_output_file_name
+                training_output_file_handler_params.AddEmptyValue("file_access_mode")
+                training_output_file_handler_params["file_access_mode"].SetString("truncate")
+                training_output_file_handler_params.AddEmptyValue("file_name")
+                training_output_file_handler_params["file_name"].SetString(training_output_file_name)
+                self.hdf5_file_training_output = KratosHDF5.HDF5FileSerial(training_output_file_handler_params)
+
+                self.hdf5_output_parameters = KM.Parameters()
+                self.hdf5_output_parameters.AddEmptyArray("list_of_variables")
+                self.hdf5_output_parameters["list_of_variables"].SetStringArray(output_var_names.GetStringArray())
+                self.hdf5_output_parameters.AddEmptyValue("prefix")
+                #self.hdf5_output_parameters["prefix"].SetString("/<time>/<model_part_name>/ResultsData")
+
+            else:
+                raise Exception("Format not supported. Supported formats are: ascii, h5.")
+
+
+    def ExecuteInitializeSolutionStep(self):
+
+        current_time = self.input_model_parts[0].ProcessInfo[KM.TIME]
+        current_step = self.input_model_parts[0].ProcessInfo[KM.STEP]
+        if((current_time >= self.interval[0]) and (current_time < self.interval[1])):
+            index = 0
+            input_value_list=[]
+            for model_part, variable, source in self.dict_input:
+                model_input_value_list = []
+                if index<len(self.perturbate_variables) and variable == self.perturbate_variables[index]:
+                    dist = self.random_distribution[index]
+                    rnd_parameters = self.random_parameters[index]
+                    factor = getattr(np.random, dist)(*rnd_parameters)
+                    # Process related variables (e.g. TIME, STEP)
+                    if source == 'process':
+                        raise Exception("Process variables cannot be perturbated.")
+                    # Node properties (e.g. position)
+                    elif source == 'node':
+                        print("Warning: The node properties are being perturbated.")
+                        for node in model_part.Nodes:
+                            input_value = op.imul(getattr(node,variable.Name()))
+                            node.SetValue(variable, input_value)
+                            model_input_value_list.append(getattr(node,variable.Name()))

```

Comment:
This whole thing can be replaced with `Kratos.Expression` (and all the similar for loops, with shared and distributed memory capabilities). I have put some parts of `Expression` in the `OptimizationApplication` thinking that it is only useful there, but now I see it is useful in here also. So first I will move them to the core, than I will tag the example usage here in this comment.

Reply:
```suggestion
                        model_input_value_expressions_list: 'list[typing.Union[KM.Expression.NodalExpression, KM.Expression.ConditionExpression, KM.ElementExpression]]' = []
                        nodal_expression = KM.Expression.NodalExpression(model_part)
                        KM.Expression.VariableExpressionIO.Read(nodal_expression, variable, False)
                        model_input_value_expressions_list.append(nodal_expression)
```

The line `KM.Expression.VariableExpressionIO.Read(nodal_expression, variable, False)` reads from the nodal non-historical since the final flag is False. If it is True, then it reads from the historical container.

To read from the condition container:
```python
    condition_expression = KM.Expression.ConditionExpression(model_part)
    KM.Expression.VariableExpressionIO.Read(condition_expression, variable)
```

If you want to get the numpy array out of these, then
```python
    numpy_array = nodal_expression.Evaluate()
```

The above code block explains how to use expressions in this case. Further examples can be found following location
https://github.com/KratosMultiphysics/Kratos/blob/39c04f39b32476e26020fda4b3eefbdfb018e934/kratos/tests/test_container_expression.py#L692-L841